llama-index-integrations/tools/llama-index-tools-azure-dynamic-sessions/llama_index/tools/azure_dynamic_sessions/base.py

```diff
@@ -0,0 +1,265 @@
+
+
+@dataclass
+class RemoteFileMetadata:
+    """Metadata for a file in the session."""
+
+    filename: str
+    """The filename relative to `/mnt/data`."""
+
+    size_in_bytes: int
+    """The size of the file in bytes."""
+
+    file_full_path: str
+    """The full path of the file."""
+
+    @staticmethod
+    def from_dict(data: dict) -> "RemoteFileMetadata":
+        """Create a RemoteFileMetadata object from a dictionary."""
+        filename = data["filename"]
+        if filename.startswith("/mnt/data"):
+            file_full_path = filename
+        else:
+            file_full_path = f"/mnt/data/{filename}"
+        return RemoteFileMetadata(
+            filename=filename,
+            size_in_bytes=data["size"],
+            file_full_path=file_full_path,
+        )
+
+
+def _sanitize_input(query: str) -> str:
+    """
+    Sanitize input and remove whitespace,
+    backtick & python (if llm mistakes python console as terminal)
+
+    Args:
+        query: The query to sanitize
+
+    Returns:
+        str: The sanitized query
+    """
+
+    # Removes `, whitespace & python from start
+    query = re.sub(r"^(\s|`)*(?i:python)?\s*", "", query)
+    # Removes whitespace & ` from end
+    query = re.sub(r"(\s|`)*$", "", query)
+    # Add new line if no new line was appended at the end of the query
+    if not query.endswith("\n"):
+        query += "\n"
+    return query
+
+
+class AzureDynamicSessionsToolSpec(BaseToolSpec):
+    """Azure Dynamic Sessions tool spec."""
+
+    spec_functions = ["code_interpreter", "list_files"]
+
+    def __init__(
+        self,
+        pool_managment_endpoint: str,
+        session_id: str = None,
+        local_save_path: str = None,
+    ) -> None:
+        """Initialize with parameters."""
+        self.pool_management_endpoint: str = pool_managment_endpoint
+
+        def _access_token_provider_factory(self) -> Callable[[], Optional[str]]:
+            self.access_token: AccessToken = None
+
+            def access_token_provider(self) -> Optional[str]:
+                """Create a function that returns an access token."""
+                if self.access_token is None or datetime.fromtimestamp(self.access_token.expires_on, timezone.utc) < (datetime.now(timezone.utc) + timedelta(minutes=5)):
+                    credential = DefaultAzureCredential()
+                    self.access_token = credential.get_token("https://dynamicsessions.io/.default")
+                return self.access_token.token
+
+            return access_token_provider
+
+        self.access_token_provider: Callable[[], Optional[str]] = _access_token_provider_factory(self)
+        """A function that returns the access token to use for the session pool."""
+
+        self.session_id: str = session_id
+        """The session ID to use for the session pool. Defaults to a random UUID."""
+        if self.session_id is None:
+            self.session_id = str(uuid4())
+
+        self.sanitize_input: bool = True
```

Comment:
Should this be configurable in the init?

Reply:
Yes, this is defaulted to `True`, but user can still set it to false if needed.